src/dist/dist.rs

```diff
@@ -701,14 +713,24 @@ fn try_update_from_dist_<'a>(
     profile: Option<Profile>,
     prefix: &InstallPrefix,
     force_update: bool,
+    components: &[&str],
+    targets: &[&str],
     fetched: &mut String,
 ) -> Result<Option<String>> {
     let toolchain_str = toolchain.to_string();
     let manifestation = Manifestation::open(prefix.clone(), toolchain.target.clone())?;
 
     // TODO: Add a notification about which manifest version is going to be used
     (download.notify_handler)(Notification::DownloadingManifest(&toolchain_str));
-    match dl_v2_manifest(download, update_hash, toolchain) {
+    match dl_v2_manifest(
+        download,
+        if components.is_empty() && targets.is_empty() {
```

Comment:
Why is this `if` necessary?

Reply:
If the toolchain exists already, and we're not explicitly trying to add components/targets then we want to skip any download if the hashes match -- this `if` blocks that short-circuit in the case that we have explicit things to add.  Without it we won't actually enter the manifestation update

Reply:
Hmm... Isn't `update_hash` only `Some` if we're doing an update, in which case `components` and `targets` will both be empty, or am I missing something?

Reply:
I decided that it was better to allow components/targets on update too, hence I needed to elide update_hash in that case.

Reply:
I guess I'm still not following how setting the `update_hash` to `None` will make a difference. Presumably it is only used to determine if a download is necessary in the first place (i.e., so that if you call `update` and the hash matches the latest manifest, it doesn't download anything). If that is the case, then setting it to `None` will only enable downloading an _older_ manifest (newer manifests will always be downloaded regardless). But `update` should never (at least in my opinion) "update" to an older nightly, even if I'm asking for a particular component or target. Beyond that, I also think that wouldn't even happen, since the backtracking code stops backtracking the moment you go beyond the current manifest date (which you would immediately).

Reply:
I believe you're correct in terms of no-backtracking as per the current algorithm.  However I think you've misunderstood the short-circuiting I was trying to describe.  In the current situation, `update`ing a channel runs thusly.

1. Download the channel manifest hash
2. *If* there is an existing hash (`update_hash` is `Some(Path)`) then compare it against the stored hash, on matching we exit "Ok, Unchanged"
3. Download the channel manifest, hash it, compare it to the downloaded hash.  If no match, we exit "Bad manifest download"
4. Continue the update with the new validated channel manifest.

What the change we are discussing does is skip 2 in the case that we're explicitly trying to add components/targets to the channel during the update.

Does that settle your concerns?  Is there something you feel I should do to make that clearer?

Reply:
Ah, I think I understand now — the `if` is there so that if the newest nightly hasn't changed, but the user wishes to add components, we do not exit early. At which point we essentially degenerate to `rustup component add`. Okay, that seems fine. I think I'd like to see a comment there saying

> Even if manifest has not changed, we must continue to install requested components.